### PR TITLE
IGNITE-17336 Spill-out to disk support for volatile RAFT log storage

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-17336

### CODE REVIEW

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/storage/impl/LocalLogStorageFactory.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+package org.apache.ignite.internal.raft.storage.impl;
+
+import org.apache.ignite.internal.raft.storage.LogStorageFactory;
+import org.apache.ignite.raft.jraft.option.RaftOptions;
+import org.apache.ignite.raft.jraft.storage.LogStorage;
+import org.apache.ignite.raft.jraft.storage.impl.LocalLogStorage;
+
+/**
+ * LogStorageFactory that always creates instances of {@link LocalLogStorage}.
```
Comment (SammyVimes) [inline modules/raft/src/main/java/org/apache/ignite/internal/raft/storage/impl/LocalLogStorageFactory.java:26]:
```suggestion
 * LogStorageFactory that creates instances of {@link LocalLogStorage}.
```

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/storage/impl/VolatileLogStorageFactoryCreator.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+        try {
+            db = RocksDB.open(this.dbOptions, this.spillOutPath.toString(), columnFamilyDescriptors, columnFamilyHandles);
+
+            // Setup rocks thread pools to utilize all the available cores as the database is shared among
+            // all the raft groups
+            Env env = db.getEnv();
+            // Setup background flushes pool
+            env.setBackgroundThreads(Runtime.getRuntime().availableProcessors(), Priority.HIGH);
+            // Setup background  compactions pool
```
Comment (SammyVimes) [inline modules/raft/src/main/java/org/apache/ignite/internal/raft/storage/impl/VolatileLogStorageFactoryCreator.java:115]:
```suggestion
            // Setup background compactions pool
```

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/storage/impl/Logs.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+     */
+    LogEntry getEntry(final long index);
+
+    /**
+     * Append entries to log.
+     */
+    void appendEntry(final LogEntry entry);
+
+    /**
+     * Append entries to log, return append success number.
```
Comment (SammyVimes) [inline modules/raft/src/main/java/org/apache/ignite/raft/jraft/storage/impl/Logs.java:40]: Probably doesn't return anything

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/storage/impl/Logs.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+     */
+    void truncatePrefix(final long firstIndexKept);
+
+    /**
+     * Delete uncommitted logs from storage's tail, (last_index_kept, last_log_index] will be discarded.
+     */
+    void truncateSuffix(final long lastIndexKept);
+
+    /**
+     * Drop all the existing logs and reset next log index to |next_log_index|. This function is called after installing
```
Comment (SammyVimes) [inline modules/raft/src/main/java/org/apache/ignite/raft/jraft/storage/impl/Logs.java:55]: I'm curious about `next_log_index`, `first_log_index`, `last_index_kept` and so on. Are these variables defined in raft paper?

**Review by SammyVimes — CHANGES_REQUESTED** [review]

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/storage/impl/Logs.java (hunk 4 above)
Comment (rpuch) [reply modules/raft/src/main/java/org/apache/ignite/raft/jraft/storage/impl/Logs.java:55]: Those are internal details of JRaft implementation, not terms from the raft paper

Comment (SammyVimes) [reply modules/raft/src/main/java/org/apache/ignite/raft/jraft/storage/impl/Logs.java:55]: Ok!

**Review by SammyVimes — APPROVED** [review]
LGTM!

**Pull request merged**